# Laboratorio 4

1. Utilice el archivo generado en el laboratorio 3 con los documentos de la colección de CACM preprocesados.
1. Obtengan el vocabulario de la colección utilizando truncamiento. Impriman la longitud del vocabulario. Después de observar el vocabulario resultante, proponga mecanismos para reducirlo. Codifique los mecanismos propuestos. Escriba la longitud del vocabulario reducido. Las longitudes deberán escribirse a pantalla
1. Escriban los vocabularios finales a archivos vocabulario.txt y vocabulario_reducido.tx



Primero extraemos las palabras de documentos_final.txt


In [1]:
filename = "documentos_final.txt"
file = open(filename, 'rt',encoding='utf-8-sig')
text = file.read()
file.close()
import re
words = re.split(r"\n", text)
doc_words = ""
i = 0
for w in words:
    split = re.split(r"\|", w)
    if len(split) == 1:
        continue
    elif len(split) == 4:
        #doc_data[i]['d_num'] = split[0]
        doc_words += split[2] + " "
    doc_words += split[1] + " "
    i = i + 1
print(doc_words[:1000])

 preliminari report intern algebra languag   extract root repeat subtract digit comput   techniqu depart matrix program scheme   glossari comput engin program terminolog  squar root approxim   comput inspect procedur   glossari comput engin program terminolog  equiv transform program scheme   propo uncol   glossari comput engin program terminolog  problem program commun chang machin propo solut part   error estim rung kutta procedur   glossari comput engin program terminolog  problem program commun chang machin propo solut part   recur curv fit techniqu   secant modif newton method   program arithmet oper   simpl automat code system   glossari comput engin program terminolog  techniqu discuss appli iter procedur solut equat accel rate converg iter converg induc converg iter diverg illustr   accel converg iter process   algebra formul flow diagram   unusu applic depart automat implement comput logic   binari truth function oper decim comput extract command   improv decim redund check   

Unimos y separamos todas las palabras (omitiendo vacías)

In [40]:
all_words = [x for x in (doc_words).split(" ") if x]
print(all_words[:100])
print(str(len(all_words)))

['preliminary', 'report', 'international', 'algebraic', 'language', 'extraction', 'root', 'repeated', 'subtraction', 'digital', 'computer', 'technique', 'department', 'matrix', 'program', 'scheme', 'glossary', 'computer', 'engineering', 'programming', 'terminology', 'square', 'root', 'approximation', 'computer', 'inspection', 'procedure', 'glossary', 'computer', 'engineering', 'programming', 'terminology', 'equivalence', 'transformation', 'program', 'scheme', 'proposal', 'uncol', 'glossary', 'computer', 'engineering', 'programming', 'terminology', 'problem', 'programming', 'communication', 'changing', 'machine', 'proposed', 'solution', 'part', 'error', 'estimation', 'runge', 'kutta', 'procedure', 'glossary', 'computer', 'engineering', 'programming', 'terminology', 'problem', 'programming', 'communication', 'changing', 'machine', 'proposed', 'solution', 'part', 'recursive', 'curve', 'fitting', 'technique', 'secant', 'modification', 'newton', 'method', 'programming', 'arithmetic', 'opera

Ahora buscaremos reducir el vocabulario existente

In [41]:
from collections import Counter

c = Counter(all_words)
#print(c.items())
labels, values = zip(*c.items())
print(labels[:10])
print(values[:10])
print(len(labels))

('preliminary', 'report', 'international', 'algebraic', 'language', 'extraction', 'root', 'repeated', 'subtraction', 'digital')
(19, 95, 7, 67, 815, 10, 66, 7, 9, 133)
7560


In [42]:
# vemos los datos en deciles
import statistics
deciles = statistics.quantiles(values, n=10)
print(deciles)

[1.0, 1.0, 1.0, 2.0, 2.0, 3.0, 6.0, 10.0, 24.0]


Viendo la distribución de los datos en deciles, vemos que la casi mitad de las palabras del vocabulario aparecen en frecuencias cercanas a 1 y 2.

Se optará por eliminar solamente las palabras con 1 ocurrencia en el vocabulario.

In [43]:
reduced_words = {k: v for k, v in c.items() if v != 1}
print(len(reduced_words))
print("We deleted " + str(len(labels) - len(reduced_words)) + " words from the vocabulary")

4712
We deleted 2848 words from the vocabulary


Vemos que se eliminaron 2848 palabras, que a mi consideración fueron demasiadas, seguiremos revisando el vocabulario existente.

Analizaremos la distribución del largo de los términos del vocabulario.

In [44]:
len_freq_dict = dict()
for w in all_words:
    len_freq_dict[len(w)] = 0
for w in all_words:
    len_freq_dict[len(w)] = len_freq_dict[len(w)] + 1
len_freq_list = (sorted(len_freq_dict.items()))
len_freq_list

[(1, 26),
 (2, 778),
 (3, 2333),
 (4, 9255),
 (5, 8942),
 (6, 12425),
 (7, 13222),
 (8, 13284),
 (9, 12396),
 (10, 7096),
 (11, 6079),
 (12, 2330),
 (13, 1336),
 (14, 998),
 (15, 191),
 (16, 133),
 (17, 23),
 (18, 11),
 (19, 1),
 (20, 3)]

In [45]:
x = len_freq_list[0][1] +len_freq_list[1][1] + len_freq_list[2][1] + len_freq_list[3][1]
print(str(x / len(all_words) * 100))

13.638264621073718


Aquí ya vemos que las palabras con 1, 2, 3 y 4 caracteres que existen en el vocabulario conforman el 13.6% del diccionario, por lo que estas serán las que se eliminarán

In [48]:
reduced_words = {k: v for k, v in c.items() if len(k) > 4}
print("We deleted " + str(len(labels) - len(reduced_words)) + " words from the vocabulary")
print(len(reduced_words))

We deleted 1091 words from the vocabulary
6469


Por último, las palabras serán escritas a vocabulario.txt y vocabulario_reducido.txt

In [47]:
voc = ', '.join(list(labels))
reduced_voc = ', '.join(list(reduced_words.keys()))
with open('vocabulario.txt', 'w') as f:
    f.write(voc)
    f.close()
with open('vocabulario_reducido.txt', 'w') as f:
    f.write(reduced_voc)
    f.close()

## Test con truncamiento
Todo el código anterior procesa un vocabulario creado con lematización, como prueba, se realizó el mismo procesamiento con un vocabulario creado a partir de truncamiento con el algoritmo de porter. Este se encuentra en los archivos **documentos_trunc.txt** y **querys_trunc.txt**

El script se ejecuta en una terminal en este directorio de la forma: 
>\>python p4.py

Los resultados fueron los siguientes:

![Resultados procesamiento vocabulario con truncamiento](test_trunc.png)

Vemos que el vocabulario inicial comienza en 5250 palabras, ¡una reducción de más de 2000 palabras en contraste con el vocabulario con lematización!

Se observa también que si buscamos eliminar palabras con 1 sola ocurrencia nos topamos con la misma situación que con lematización, el vocabulario se reduciría demasiado (más de la mitad), por lo que la opción de eliminar por largo será la más adecuada en este caso.

Podemos solamente omitr palabras de largo = {1,2,3} y no de 4 también, eliminando el 9.85% del vocabulario de esta forma y teniendo un vocabulario final de 4670 términos.